sympy/combinatorics/free_group.py

```diff
@@ -0,0 +1,918 @@
+                    "`FreeGroupElm`")
+        return self.contains(i)
+
+    def __len__(self):
+        return self.rank
+
+    def __str__(self):
+        if self.rank > 30:
+            str_form = "<free group with %s generators>" % self.rank
+        else:
+            str_form = "<free group on the generators "
+            gens = self.generators
+            str_form += str(gens) + ">"
+        return str_form
+
+    __repr__ = __str__
+
+    def __eq__(self, other):
+        """No ``FreeGroup`` is equal to any "other" ``FreeGroup``.
+
+        Examples
+        ========
+
+        >>> from sympy import FreeGroup
+        >>> f = FreeGroup( 4, "swapnil" )
+        >>> g = FreeGroup( 4, "swapnil" )
+        >>> f == g
+        False
+        >>> f == f
+        True
+
+        """
+        if self is other:
+            return True
+        else:
+            return False
+
+    @property
+    def as_str(self):
+        return self._str
+
+    def order(self):
+        if self.rank == 0:
+            return 1
+        else:
+            return S.Infinity
+
+    @property
+    def elements(self):
+        if self.rank == 0:
+            # A set containing Identity element of `FreeGroup` self is returned
+            return FreeGroupElm(self, [])
+        else:
+            raise ValueError("Group contains infinitely many elements"
+                            ", hence can't be represented")
+
+    @property
+    def rank(self):
+        r"""
+        In group theory, the `rank` of a group `G`, denoted `G.rank`,
+        can refer to the smallest cardinality of a generating set
+        for G, that is
+
+        \operatorname{rank}(G)=\min\{ |X|: X\subseteq G, \langle X\rangle =G\}.
+
+        """
+        return self._rank
+
+    @property
+    def is_abelian(self):
+        """Test if the group is Abelian.
+
+        Examples
+        ========
+
+        >>> from sympy import FreeGroup
+        >>> f = FreeGroup( 4 )
+        >>> f.is_abelian
+        False
+
+        >>> g = FreeGroup( 0 )
+        >>> g.is_abelian
+        True
+
+        """
+        if self.rank == 0:
+            return True
+        else:
+            return False
+
+    def contains(self, g):
+        """Test if Free Group element ``g`` belong to self, ``G``.
+
+        In mathematical terms any linear combination of generators
+        of a Free Group is contained in it.
+
+        Examples
+        ========
+
+        >>> from sympy import FreeGroup
+        >>> f = FreeGroup( 4 )
+        >>> f.contains(f[0]**3*f[1]**2)
+        True
+
+        """
+        if not isinstance(i, FreeGroupElm):
+            return False
+        elif self != i.group:
+            return False
+        else:
+            return True
+
+    def is_subgroup(self, F):
+        """Return True if all elements of `self` belong to `F`.
+
+        Examples
+        ========
+        """
+        if F.is_group and set(F.generators).issubset(set(self.generators)):
```

Comment:
I think this should be something like ``all([self.contains(gen) for gen in F.generators])``.